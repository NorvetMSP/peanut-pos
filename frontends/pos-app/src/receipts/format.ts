import type { PrintBlock, PrintJob } from "../devices/printer";
import type { CartItem } from "../CartContext";

export type SaleReceipt = {
  orderId?: string;
  storeLabel: string;
  cashierLabel: string;
  items: CartItem[];
  subtotal: number;
  tax?: number;
  total: number;
  paidMethod: string;
  createdAt: Date;
  footerNote?: string;
};

const fmtMoney = (n: number) => `$${n.toFixed(2)}`;

function line(text: string, width = 42, align: "left" | "center" | "right" = "left"): PrintBlock {
  const content = text.length > width ? text.slice(0, width) : text;
  return { type: "text", content, align };
}

function hr(width = 42): PrintBlock {
  return line("-".repeat(width));
}

export function buildSaleReceiptJob(data: SaleReceipt, width = 42): PrintJob {
  const blocks: PrintBlock[] = [];

  // Header
  blocks.push({ type: "text", content: "NovaPOS", align: "center", bold: true, size: "m" });
  blocks.push(line(data.storeLabel, width, "center"));
  blocks.push(line(`Cashier: ${data.cashierLabel}`, width, "center"));
  const ts = data.createdAt.toLocaleString();
  blocks.push(line(ts, width, "center"));
  if (data.orderId) blocks.push(line(`Order # ${data.orderId}`, width, "center"));
  blocks.push(hr(width));

  // Items
  data.items.forEach((it) => {
    const name = it.name.length > width ? it.name.slice(0, width - 1) : it.name;
    blocks.push(line(name, width, "left"));
    const qtyPrice = `x${it.quantity} @ ${fmtMoney(it.price)}`;
    const lineTotal = fmtMoney(it.price * it.quantity);
    const pad = Math.max(1, width - qtyPrice.length - lineTotal.length);
    blocks.push(line(`${qtyPrice}${" ".repeat(pad)}${lineTotal}`, width, "left"));
  });

  blocks.push(hr(width));

  // Totals
  const subtotal = data.subtotal;
  const tax = data.tax ?? Math.max(0, data.total - subtotal);
  const rows: Array<[string, number]> = [
    ["Subtotal", subtotal],
    ["Tax", tax],
    ["Total", data.total],
  ];
  for (const [label, val] of rows) {
    const left = label;
    const right = fmtMoney(val);
    const pad = Math.max(1, width - left.length - right.length);
    blocks.push(line(`${left}${" ".repeat(pad)}${right}`, width));
  }

  blocks.push(line(`Paid: ${data.paidMethod.toUpperCase()}`, width));
  blocks.push(hr(width));
  blocks.push(line("Thank you!", width, "center"));
  if (data.footerNote) blocks.push(line(data.footerNote, width, "center"));

  return { widthChars: width, blocks, cut: true };
}
