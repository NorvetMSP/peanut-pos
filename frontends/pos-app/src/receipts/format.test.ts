import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { buildSaleReceiptJob, type SaleReceipt } from './format';

const realToLocaleString = Date.prototype.toLocaleString;

describe('buildSaleReceiptJob', () => {
  beforeAll(() => {
    // Stabilize date formatting for snapshot consistency
    // eslint-disable-next-line no-extend-native
    Date.prototype.toLocaleString = function (this: Date) {
      const y = this.getUTCFullYear();
      const m = String(this.getUTCMonth() + 1).padStart(2, '0');
      const d = String(this.getUTCDate()).padStart(2, '0');
      const hh = String(this.getUTCHours()).padStart(2, '0');
      const mm = String(this.getUTCMinutes()).padStart(2, '0');
      const ss = String(this.getUTCSeconds()).padStart(2, '0');
      return `${y}-${m}-${d} ${hh}:${mm}:${ss} UTC`;
    } as any;
  });

  afterAll(() => {
    Date.prototype.toLocaleString = realToLocaleString;
  });

  const sample: SaleReceipt = {
    orderId: '11111111-1111-1111-1111-111111111111',
    storeLabel: 'Flagship Store',
    cashierLabel: 'Casey Cashier',
    items: [
      { id: 'p1', name: 'Premium Coffee Beans 1kg', price: 14.99, quantity: 1, sku: 'COF-1KG' },
      { id: 'p2', name: 'Reusable Cup', price: 9.5, quantity: 2, sku: 'CUP-R' },
    ],
    subtotal: 14.99 + 9.5 * 2,
    total: 14.99 + 9.5 * 2 + 2.34,
    tax: 2.34,
    paidMethod: 'card',
    createdAt: new Date(Date.UTC(2025, 0, 2, 3, 4, 5)),
    footerNote: 'Thanks for shopping!',
  };

  it('builds a job at width 42', () => {
    const job = buildSaleReceiptJob(sample, 42);
    expect(job.widthChars).toBe(42);
    expect(job.blocks.length).toBeGreaterThan(0);
    expect(job).toMatchSnapshot();
  });

  it('builds a job at width 32', () => {
    const job = buildSaleReceiptJob(sample, 32);
    expect(job.widthChars).toBe(32);
    expect(job.blocks.length).toBeGreaterThan(0);
    expect(job).toMatchSnapshot();
  });
});
