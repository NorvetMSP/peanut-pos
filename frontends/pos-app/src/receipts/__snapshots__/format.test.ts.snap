// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildSaleReceiptJob > builds a job at width 32 1`] = `
{
  "blocks": [
    {
      "align": "center",
      "bold": true,
      "content": "NovaPOS",
      "size": "m",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Flagship Store",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Cashier: Casey Cashier",
      "type": "text",
    },
    {
      "align": "center",
      "content": "2025-01-02 03:04:05 UTC",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Order # 11111111-1111-1111-1111-",
      "type": "text",
    },
    {
      "align": "left",
      "content": "--------------------------------",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Premium Coffee Beans 1kg",
      "type": "text",
    },
    {
      "align": "left",
      "content": "x1 @ $14.99               $14.99",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Reusable Cup",
      "type": "text",
    },
    {
      "align": "left",
      "content": "x2 @ $9.50                $19.00",
      "type": "text",
    },
    {
      "align": "left",
      "content": "--------------------------------",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Subtotal                  $33.99",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Tax                        $2.34",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Total                     $36.33",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Paid: CARD",
      "type": "text",
    },
    {
      "align": "left",
      "content": "--------------------------------",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Thank you!",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Thanks for shopping!",
      "type": "text",
    },
  ],
  "cut": true,
  "widthChars": 32,
}
`;

exports[`buildSaleReceiptJob > builds a job at width 42 1`] = `
{
  "blocks": [
    {
      "align": "center",
      "bold": true,
      "content": "NovaPOS",
      "size": "m",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Flagship Store",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Cashier: Casey Cashier",
      "type": "text",
    },
    {
      "align": "center",
      "content": "2025-01-02 03:04:05 UTC",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Order # 11111111-1111-1111-1111-1111111111",
      "type": "text",
    },
    {
      "align": "left",
      "content": "------------------------------------------",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Premium Coffee Beans 1kg",
      "type": "text",
    },
    {
      "align": "left",
      "content": "x1 @ $14.99                         $14.99",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Reusable Cup",
      "type": "text",
    },
    {
      "align": "left",
      "content": "x2 @ $9.50                          $19.00",
      "type": "text",
    },
    {
      "align": "left",
      "content": "------------------------------------------",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Subtotal                            $33.99",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Tax                                  $2.34",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Total                               $36.33",
      "type": "text",
    },
    {
      "align": "left",
      "content": "Paid: CARD",
      "type": "text",
    },
    {
      "align": "left",
      "content": "------------------------------------------",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Thank you!",
      "type": "text",
    },
    {
      "align": "center",
      "content": "Thanks for shopping!",
      "type": "text",
    },
  ],
  "cut": true,
  "widthChars": 42,
}
`;
